### PR TITLE
[FLINK-3521] Make Iterable part of method signature for WindowFunction

### PR DESCRIPTION
**Pull request by aljoscha**

### CODE REVIEW

**Comment by rmetzger** [comment]
Scrolled over the changes, didn't spot anything suspicious

**Comment by StephanEwen** [comment]
Looks good, but I need a few more minutes to check this...

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/function/AllWindowFunction.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.io.Serializable
+
+import org.apache.flink.annotation.Public
+import org.apache.flink.api.common.functions.Function
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.util.Collector
+
+/**
+  * Base interface for functions that are evaluated over keyed (grouped) windows.
```
Comment (StephanEwen) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/function/AllWindowFunction.scala]: JavaDoc is wrong here

**Comment by StephanEwen** [comment]
May be unrelated to this change, but I think `fold()` for non-keyed windows is not incremental, meaning it has no constant space requirement.

**Comment by aljoscha** [comment]
Yes, this is correct. (about the fold() on non-keyed windows)

**Comment by StephanEwen** [comment]
I like this, especially the natural Scala path. Some concerns, though:

  1. As noted before, the fold function in the non-keyed window is not space constant. That is quite a serious issue, if people count heavy streams in large windows. That should be fixed.

  2. The `reduce(...)` code on the `WindowedStream` creates a pass-through window function over Iterable, instead of an pass-through internal window function. The result is that for reduced windows, the single result value is wrapped and unwrapped, instead of simply passed through. That should be a quite simple fix.

  3. Minor comment, more for future code, (and that applies to virtually all Scala API code): The code creates function objects in line a lot. As a result, we have for example at least 10 classes that implement a Flink ReduceFunction and call a Scala Function2. I would suggest to define that class once externally (not inline), because it reduces number of subclasses of the reduce function and that may actually have a positive impact on the JIT friendliness of some programs (if the number of loaded implementations of ReduceFunction at any point becomes reasonably small)

**Comment by aljoscha** [comment]
The `PassThroughFunction` is created because at that point it is not yet clear what type of `InternalWindowFunction` will be created. I can fix it by duplicating all the WindowOperator creation code, but this applies to reduce/fold both on the Java and Scala API.

**Comment by aljoscha** [comment]
1. Is very hard to fix, ditching the non-keyed window operators altogether and using a dummy key selector would be easier.

**Comment by StephanEwen** [comment]
I think (2) can be done easier if the operator creation code goes into something more generic, but that is not that urgent.

**Comment by StephanEwen** [comment]
Let's merge this and open a follow-up issue for (1)

**Comment by rmetzger** [comment]
Merging ...

**Pull request closed without merging**